Update instruction logging for RISCV

qwattash opened the pull request:
This brings the same functionality for instruction tracing that we currently have on the CHERI-MIPS target.

qwattash committed:
riscv: Implement register (GPR, CSR and capability) and instruction logging.

- Move instruction logging helpers to op_helper_log_instr.c to mirror MIPS
code layout.
- Add instrumentation to GPR and CSR register updates.
- Most CHERI register updates should come for free from cheri-common
logging instrumentation.
- Add logging for RISCV ASID.

qwattash committed:
riscv: Implement magic nops to control tracing.

- slti zero, zero, 0x1b => begin instruction tracing
- slti zero, zero, 0x1e => end instruction tracing
- slti zero, zero, 0x2b => begin user-only instruction tracing
- slti zero, zero, 0x2e => end user-only instruction tracing

qwattash committed:
Make sure we commit to the log the instruction that disables instruction logging.

qwattash committed:
cheri: Rename log_changed_capreg_* variants.

These are renamed to cheri_log_instr_changed_capreg_* to better
hint that are logging helpers defined in the common cheri layer.

qwattash commented:
@heshamelmatary This is another good suggestion, probably better than reserving an opcode. Moving the functionality around is quite easy so I'll defer this after I have the rest of the patch working.

heshamelmatary commented:
@qwattash Yeah I didn't mean it for this PR. Just a suggestion for a future enhancement.

qwattash committed:
riscv: Add instruction logging for CSR and CSR registers and exceptions.

- Add helpers to emit register contents into the current instruction
  extra debug message buffer.
- Add placeholders for missing VS-* capability SCRs.
- Add logging for CSRs changed when swapping state between H and S modes.
- Add logging for CSRs changed on exception and mret/sret.
- Add logging exceptions and interrupts.

Currently this does not log changes to *_hs background registers because
it is unclear how to fetch their name.

qwattash committed:
Extend implementation of user-only logging.

We now use the qemu_log_instr_mode_switch API to log a generic change in
CPU execution mode. This can be used to enable user-only logging but also
other forms of filters based on the CPU mode, and allows inclusion of the
mode into traces.

When we stop logging, we delay commit of the last instruction to the
beginning of the next TB. This is because in case we want to stop logging
on a return to lower EL, we need to wait to capture possible interrupt delivery
that happens upon returning.

qwattash committed:
riscv: Replace old extra instruction logging output.

arichardson commented on target/riscv/Makefile.objs:
```diff
@@ -1,6 +1,7 @@
 obj-y += translate.o op_helper.o cpu_helper.o cpu.o csr.o fpu_helper.o gdbstub.o
 obj-$(CONFIG_SOFTMMU) += pmp.o
 obj-$(TARGET_CHERI) += op_helper_cheri.o
+obj-${CONFIG_TCG_LOG_INSTR} += op_helper_log_instr.o
```
Should use parens to match other lines.

arichardson commented on include/qemu/log_instr.h:
```diff
@@ -67,6 +85,10 @@ struct cpu_log_instr_info;
 typedef struct {
     /* CPU tracing user mode only enable flag */
     bool user_mode_tracing;
+    /* /\* Are we switching mode? *\/ */
```
Are you planning to uncomment these in a follow up commit?

qwattash replied:
I meant to delete them but forgot about it, thanks for pointing it out.

arichardson commented on target/riscv/cpu_helper.c:
```diff
@@ -113,7 +114,61 @@ bool riscv_cpu_fp_enabled(CPURISCVState *env)
+    CPURISCVState *env, bool current_virt)
+{
+    if (current_virt) {
+        /* Current V=1 and we are about to change to V=0 */
```
Could this be de-deduplicated using macros like the COPY_SPECIAL? I'm also not sure factoring this out is a good idea since we won't get merge conflicts when the hypervisor stuff changes. I assume it will change a bit in the future since it's still very new.

qwattash replied:
I can de-duplicate it using COPY_SPECIAL approach, I opted to do it this way to avoid having COPY_SPECIAL with tons of parameters, since I would have had to add the CSR_* number and the CheriSCR_* number as well. I will revert back to inlining these as getting merge conflicts for hypervisor changes will definitely help during merges.

arichardson commented on target/riscv/insn_trans/trans_rvi.inc.c:
```diff
@@ -278,6 +278,43 @@ static void gen_sltu(TCGv ret, TCGv s1, TCGv s2)
 {
+#ifdef CONFIG_TCG_LOG_INSTR
+    /*
+     * If instruction tracing is enabled, we use addi zero, zero, <magic>
```
Slti?

qwattash replied:
I originally used addi, however @jrtc27 correctly pointed out that I was using a reserved opcode for HINT and suggested this alternative that is not compressible and not reserved.

arichardson replied:
Yeah I just meant the comment is stale
```suggestion
     * If instruction tracing is enabled, we use slti zero, zero, <magic>
```

arichardson commented on target/riscv/cheri-archspecific-early.h:
```diff
@@ -84,6 +84,16 @@ enum CheriSCR {
+    CheriSCR_BSTDC = 33,
+    CheriSCR_BSScratchC = 34,
+    CheriSCR_BSEPCC = 35,
+    CheriSCR_MAX,
```
Can we have more than 32? I'm not sure what the instruction format is.

qwattash replied:
I assume not, which is why I commented it with the TODO, specifiying that these are just placeholders. We currently don't have the instruction to access them implemented.

arichardson reviewed (approved):
Looks fine to me if it works for you. Just a few minor comments.

arichardson commented on target/riscv/cpu_helper.c:
```diff
@@ -1062,41 +1154,46 @@ void riscv_cpu_do_interrupt(CPUState *cs)
+        riscv_log_instr_csr_changed(env, CSR_SCAUSE);
+#ifdef TARGET_CHERI
+        env->SEPCC = env->PCC;
+        riscv_log_instr_scr_changed(env, CheriSCR_SEPCC);
```
Can we include that in the COPY_SPECIAL_REG macro?

qwattash replied:
Good point

qwattash committed:
riscv: replace more CSR/SCR copies with helper macro.

arichardson commented:
@qwattash The PR doesn't build for x86. Could you add the necessary stubs? And maybe also for aarch64 if you can be bothered.

qwattash commented:
@arichardson I think we should disable tcg_log_instr by default in the qemu configure script and enable it from cheribuild for the targets that support it.

qwattash committed:
Only enable instruction tracing for targets that support it.
